File: Data Structures & Algorithms/sliding-window-maximum/submission-1.cs
public class Solution {
    public int[] MaxSlidingWindow(int[] nums, int k) {
        // priority queue in c# is min heap by default 
        // so we need to negative the priority when push to heap
        var heap = new PriorityQueue<(int value, int index), int>();
        var result = new List<int>();

        for (int i = 0; i < nums.Length; i++) {
            // add the current number to heap
            heap.Enqueue((nums[i], i), -nums[i]);

            // remove max number if it is out of window range
            while (heap.Peek().index <= i - k) heap.Dequeue();

            if (i - k + 1 >= 0) result.Add(heap.Peek().value);
        }

        return result.ToArray();
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-1.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        std::vector<std::vector<int>> result{};
        std::vector<int> subset{};
        backtrack(nums, 0, result, subset);
        return result;
    }

    void backtrack(std::vector<int>& nums, int index, std::vector<std::vector<int>>& result,
        std::vector<int>& subset) {
            if (index == nums.size()) {
                result.push_back(subset);
                return;
            }

            int current = nums[index];
            subset.push_back(current);
            backtrack(nums, index + 1, result, subset);
            subset.pop_back();

            while (current == nums[index]) index++;
            backtrack(nums, index, result, subset);
    }
};
